### PR TITLE
Teams ActionBasedMessagingExtensionFetchTask Scenario

### PR DESCRIPTION
**Pull request by EricDahlvang**
Add MessagingExtensionQuery param to OnTeamsMessagingExtensionFetchTaskAsync

Update ActionBasedMessagingExtensionFetchTaskBot and Readme.md

### CODE REVIEW

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

File: tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs [hunk 1]
```diff
@@ -17,5 +20,110 @@ protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivi
...
+                        Title = "Task Module Example",
+                    },
+                },
+            };
+        }
+
+        protected override async Task<MessagingExtensionActionResponse> OnTeamsMessagingExtensionSubmitActionAsync(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionAction action, CancellationToken cancellationToken)
+        {
+            var reply = MessageFactory.Text("OnTeamsMessagingExtensionSubmitActionAsync MessagingExtensionAction: " + JsonConvert.SerializeObject(action));
+            await turnContext.SendActivityAsync(reply, cancellationToken);
```
Comment (stevengum) [inline tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs]: Does this work if I use the action extension in a channel that the bot does not have access to?

Part of me thinks this scenario is fine, and another part of me thinks we should perhaps break out the `OnTeamsMessagingExtensionSubmitActionAsync` into a separate Scenario.

File: tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs [hunk 2]
```diff
@@ -17,5 +20,110 @@ protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivi
...
+
+            var adaptiveCard = GetCardFromSubmitExampleData(JsonConvert.DeserializeObject<SubmitExampleData>(action.Data.ToString()));
+
+            return new MessagingExtensionActionResponse
```
Comment (stevengum) [inline tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs]: Is this behavior in a screenshot in the README? I don't think I saw it there...

File: tests/Teams/ActionBasedMessagingExtension-FetchTask/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  ```bash
+  # determine dotnet version
+  dotnet --version
+  ```
+
+### Clone the repository
+
+```bash
+git clone https://github.com/Microsoft/botbuilder-samples.git
```
[the remark this answers is not in the document]
Comment (EricDahlvang) [reply tests/Teams/ActionBasedMessagingExtension-FetchTask/README.md]: fixed

File: tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs (hunk 1 above)
Comment (EricDahlvang) [reply tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs]: No, sending to a conversation in which the bot is not installed throws an Unauthorized.  

Do you think we should either 1) remove these lines that send the message, or 2) add a try/catch and ignore the exception ?

I think having OnTeamsMessagingExtensionSubmitActionAsync in this Scenario helps differentiate it from the non-task module Messaging Extension Scenario  But, I don't feel strongly about this.  We should discuss which Scenarios we want to cover which functionality (there is a multitude of different flows possible, and the interactions vary depending on a variety of factors).

File: tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs (hunk 2 above)
Comment (EricDahlvang) [reply tests/Teams/ActionBasedMessagingExtension-FetchTask/Bots/ActionBasedMessagingExtensionFetchTaskBot.cs]: Step 4?

File: tests/Teams/ActionBasedMessagingExtension-FetchTask/README.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+20. Click "Add" 
+21. Copy the value and put it into Notepad. Label it "Password"
+22. (back in the channel registration view) Copy/Paste the Microsoft App ID and Password into their respective fields 
+23. Click Create 
+24. Go to "Resource groups" on the left 
+25. Select the resource group that the bot channel reg was created in 
+26. Select the bot channel registration 
+27. Go to Settings  
+28. Select the "Teams" icon under "Add a featured channel 
+29. Click Save
```
[the remark this answers is not in the document]
Comment (EricDahlvang) [reply tests/Teams/ActionBasedMessagingExtension-FetchTask/README.md]: I copied the steps from a different Readme.

Maybe we can just link to here: https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/bots/bots-create#create-a-bot-for-microsoft-teams  (note: the steps in this doc don't require an Azure Subscription)

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.3.1)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.3.1](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.3.1)

File: tests/Teams/ActionBasedMessagingExtension-FetchTask/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnataylor) [reply tests/Teams/ActionBasedMessagingExtension-FetchTask/README.md]: Tien added a bunch of detail in the Readme for unfurling. I think for the moment its OK to have the duplication. But the most useful thing for me is where we talk about Teams.

File: tests/Microsoft.Bot.Builder.Teams.Tests/TeamsActivityHandlerTests.cs [hunk 5]
```diff
@@ -793,10 +794,10 @@ protected override Task OnTeamsMessagingExtensionCardButtonClickedAsync(ITurnCon
                 return base.OnTeamsMessagingExtensionCardButtonClickedAsync(turnContext, cancellationToken);
             }
 
-            protected override Task<MessagingExtensionActionResponse> OnTeamsMessagingExtensionFetchTaskAsync(ITurnContext<IInvokeActivity> turnContext, CancellationToken cancellationToken)
+            protected override Task<MessagingExtensionActionResponse> OnTeamsMessagingExtensionFetchTaskAsync(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionQuery query, CancellationToken cancellationToken)
             {
                 Record.Add(MethodBase.GetCurrentMethod().Name);
-                return base.OnTeamsMessagingExtensionFetchTaskAsync(turnContext, cancellationToken);
+                return base.OnTeamsMessagingExtensionFetchTaskAsync(turnContext, query, cancellationToken);
```
Comment (johnataylor) [inline tests/Microsoft.Bot.Builder.Teams.Tests/TeamsActivityHandlerTests.cs]: my bad... these tests could perhaps be a little deeper :-(
(they only check the dispatching not the values in the query)

**Pull request merged**